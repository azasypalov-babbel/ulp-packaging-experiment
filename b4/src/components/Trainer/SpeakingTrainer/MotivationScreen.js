import React from 'react';
import PropTypes from 'prop-types';
import Text from '../../shared/Text';
import styled from 'styled-components';
import Button from '../../shared/Button/Button';
import CenterContent from '../../shared/CenterContent';

import motivationImage from '@assets/images/speakingTrainer/feel-silly-motivation.svg';
import withTranslations from '../../shared/withTranslations';

const Layout = styled.div`
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  max-width: 35rem;
  & > * {
    margin: 1.5rem 0;
  }
`;

const MotivationScreen = ({ onFinish, translations }) => (
  <CenterContent>
    <Layout>
      <img src={motivationImage} />
      <Text
        as="h1"
        textAlign="center"
        fontSize={'large'}
        color={'spaceGrayW15'}
        fontFamily={'fontLeituraBold'}>
        {translations.title}
      </Text>
      <Text textAlign="center">
        {translations.body}
      </Text>
      <Button data-selector="got-it" primary onClick={onFinish}>{translations.button}</Button>
    </Layout>
  </CenterContent>
);

MotivationScreen.propTypes = {
  onFinish: PropTypes.func,
  translations: PropTypes.shape({
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired,
    button: PropTypes.string.isRequired
  }).isRequired
};

const getTranslations = (translate) => ({
  title: translate('speech_recognition.feeling_silly.title'),
  body: translate('speech_recognition.feeling_silly.body'),
  button: translate('speech_recognition.got_it')
});

export default withTranslations(getTranslations)(MotivationScreen);
