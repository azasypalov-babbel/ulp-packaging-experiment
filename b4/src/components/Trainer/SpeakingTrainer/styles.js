import styled from 'styled-components';
import { rem } from 'polished';

export const StyledWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;

  max-width: ${rem(732)};
  margin: ${rem(40)} auto;

  box-sizing: border-box;
`;

