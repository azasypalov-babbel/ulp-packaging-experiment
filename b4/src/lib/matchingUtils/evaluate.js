
import levenshteinDistance from './levenshteinDistance';

const averageDistance = (stringA, stringB, distance) =>
  distance / Math.max(stringA.length, stringB.length);

export const getSubstitutionCost = (characterSubstitutionCosts = []) => (a, b) => {
  if (a === b) return 0;

  const substitution = characterSubstitutionCosts
    .find(({ character, replacement }) => a === character && b === replacement);
  if (substitution) {
    return substitution.cost;
  }

  return 1;
};

export const RESULTS = {
  CORRECT: 'CORRECT',
  INCORRECT: 'INCORRECT',
  TYPO: 'TYPO'
};

const TYPO_THRESHOLD = 0.25;

export const evaluate = ({ attemptText, targetText, substitutionCosts = [] }) => {
  if (attemptText === targetText) {
    return RESULTS.CORRECT;
  }

  const targetTokens = targetText.split(/\s+/);
  const attemptTokens = attemptText.split(/\s+/);
  if (targetTokens.length != attemptTokens.length) {
    return RESULTS.INCORRECT;
  }
  let hasTypo = false;
  for (let i = 0; i < targetTokens.length; i += 1) {
    const distance = levenshteinDistance(
      targetTokens[i],
      attemptTokens[i],
      getSubstitutionCost(substitutionCosts)
    ).distance;
    if (averageDistance(targetTokens[i], attemptTokens[i], distance) > TYPO_THRESHOLD) {
      return RESULTS.INCORRECT;
    } else if (distance > 0) {
      hasTypo = true;
    }
  }
  return hasTypo ? RESULTS.TYPO : RESULTS.CORRECT;
};
