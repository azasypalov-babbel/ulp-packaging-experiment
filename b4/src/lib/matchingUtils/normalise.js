import {
  removeLeadingAndTrailingPunctuation,
  createSpecialCharacterReplacements
} from '@lessonnine/babbel-markup-helper.js';

const removeTags = (string) => string.replace(/[#@]/g, '');
const removePunctuation = (string) => string.replace(/[,;:./]/g, '');
const removeUnicodeEllipses = (string) => string.replace(/\u2026/g, '');
const removeDashes = (string) => string.replace(/[－﹣—–‑‐-]\s/g, '');
const removeParentheses = (string) => string.replace(/\([^)]*\)\s/g, '');
const removeExtraSpaces = (string) => string.replace(/\s+/g, ' ');
const stardardString = /^[a-zA-Z0-9]+$/;


const normalisers = [
  removeLeadingAndTrailingPunctuation,
  removeTags,
  removePunctuation,
  removeUnicodeEllipses,
  removeDashes,
  removeParentheses,
  removeExtraSpaces
];

export const normalise = (string, learnLanguageAlpha3) =>
  (learnLanguageAlpha3 === 'DEU') ?
    normalisers
      .reduce((result, normaliser) => normaliser(result), string)
      .trim()
    :
    normalisers
      .reduce((result, normaliser) => normaliser(result), string)
      .trim()
      .toLowerCase();

export const replaceSpecialCharacters = (char) => {
  if (char.match(stardardString)) {
    return char;
  }

  const specialChars = createSpecialCharacterReplacements(char);
  if (!specialChars.length) return char;

  return specialChars
    .filter((string) => string.match(stardardString))
    .find((string) => string.length == 1);
};

export const replaceMatches = (inputString, replacementPairs = []) =>
  replacementPairs.reduce((string, { expression, replacement }) =>
    string.replace(new RegExp(expression, 'ig'), replacement), inputString);

export const stripDiacritics = (string) => string.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
