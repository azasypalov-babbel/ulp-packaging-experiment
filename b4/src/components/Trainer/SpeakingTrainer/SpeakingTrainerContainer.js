import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { formatDL } from '../../../lib/markupFormatter';
import playFeedbackSound, { FEEDBACK_SOUND } from '../../../lib/playFeedbackSound';
import {
  formatParentheses,
  getFirstCorrectSolution,
  removeGapFormatting,
  markupStringToPlainText,
  stripCurlyBracesAndContent
} from '@lessonnine/babbel-markup-helper.js';

import * as sequenceActions from '../../../dux/sequence/actions';
import { track } from '../../../dux/tracker/actions';

import withAudio, { audioPropTypes } from '../../shared/withAudio';
import withSpeech, {
  speechPropTypes,
  recognitionHasEnded,
  recognitionHasError
} from '../../shared/withSpeech';
import { withServices } from '../../shared/withServices';

import { learnLanguageAlpha3toBCP47 } from '../../../lib/languageCodeConverter';
import SpeakingTrainer from './SpeakingTrainer';
import MotivationScreen from './MotivationScreen';

import withSpeechMessages from '../../shared/withSpeechMessages';
import { getLocalStorage } from '../../../lib/localStorage';


const stripParantheses = (text) => text.replace(/(\([\S]+\))/g, '').trim();

const formatLL = compose(
  stripParantheses,
  removeGapFormatting,
  formatParentheses,
  markupStringToPlainText,
  getFirstCorrectSolution,
  stripCurlyBracesAndContent,
  (s) => s || ''
);

const formatTrainerItem = (item) => ({
  id: item.id,
  sound: item.sound,
  learnLanguageText: formatLL(item.learn_language_text),
  displayLanguageText: formatDL(item.display_language_text)
});

const flattenItems = (trainer) => {
  return trainer.item_groups.reduce((acc, group) => {
    return acc.concat(group.items);
  }, []);
};

const SPEAKING_TRAINER_ONBOARDING_COMPLETE_KEY = 'SPEAKING_TRAINER_ONBOARDING_COMPLETE';
const LOCALSTORAGE_NAMESPACE = 'SPEAKING';
const localStorage = getLocalStorage(LOCALSTORAGE_NAMESPACE);

export class SpeakingTrainerContainer extends React.Component {
  constructor(props) {
    super(props);

    const items = flattenItems(props.trainer).map(formatTrainerItem);

    this.state = {
      items,
      attempt: undefined,
      currentIndex: 0,
      userAudioUrl: null,
      onboardingCompleted: localStorage.get(SPEAKING_TRAINER_ONBOARDING_COMPLETE_KEY) === true
    };

    this.handleContinueClick = this.handleContinueClick.bind(this);
    this.handleAttempt = this.handleAttempt.bind(this);
    this.handleSpeechEnd = this.handleSpeechEnd.bind(this);
    this.handleSpeechError = this.handleSpeechError.bind(this);
    this.startSpeech = this.startSpeech.bind(this);
    this.handleAttemptRetry = this.handleAttemptRetry.bind(this);
    this.handleMicButtonClick = this.handleMicButtonClick.bind(this);
    this.handleFeedbackAttemptCardClick = this.handleFeedbackAttemptCardClick.bind(this);
    this.handleItemPlaybackCardClick = this.handleItemPlaybackCardClick.bind(this);
    this.handleOnboardingComplete = this.handleOnboardingComplete.bind(this);
    this.handleSoundPlayEnded = this.handleSoundPlayEnded.bind(this);
  }

  UNSAFE_componentWillMount() {
    const { items } = this.state;
    this.props.onStart({ scorableItemsCount: items.length });
  }

  handleFeedbackAttemptCardClick() {
    if (!this.state.onboardingComplete && this.state.currentIndex === 0) {
      const { track, locale, learnLanguageAlpha3 } = this.props;

      /* eslint-disable camelcase */
      track({
        event: 'gui:interacted',
        version: 1,
        payload: {
          gui_container: 'speaking_trainer',
          gui_element: 'feedback_attempt_card',
          interaction: 'clicked',
          origin: 'lesson_player',
          locale: locale,
          learn_language_alpha3: learnLanguageAlpha3
        }
      });
      /* eslint-enable camelcase */
    }
  }

  componentDidUpdate(prevProps) {
    if (recognitionHasEnded(prevProps.speech, this.props.speech)) this.handleSpeechEnd();
    if (recognitionHasError(prevProps.speech, this.props.speech)) this.handleSpeechError();
  }

  handleItemPlaybackCardClick() {
    if (!this.state.onboardingComplete && this.state.currentIndex === 0) {
      const { track, locale, learnLanguageAlpha3 } = this.props;

      /* eslint-disable camelcase */
      track({
        event: 'gui:interacted',
        version: 1,
        payload: {
          gui_container: 'speaking_trainer',
          gui_element: 'item_playback_card',
          interaction: 'clicked',
          origin: 'lesson_player',
          locale: locale,
          learn_language_alpha3: learnLanguageAlpha3
        }
      });
      /* eslint-enable camelcase */
    }
  }

  handleMicButtonClick() {
    this.startSpeech();

    if (!this.state.onboardingCompleted && this.state.currentIndex === 0) {
      const { track, locale, learnLanguageAlpha3 } = this.props;

      /* eslint-disable camelcase */
      track({
        event: 'gui:interacted',
        version: 1,
        payload: {
          gui_container: 'speaking_trainer',
          gui_element: 'mic_button',
          interaction: 'clicked',
          origin: 'lesson_player',
          locale: locale,
          learn_language_alpha3: learnLanguageAlpha3
        }
      });
      /* eslint-enable camelcase */
    }
  }

  handleSoundPlayEnded() {
    this.startSpeech();
  }

  handleAttemptRetry() {
    this.props.speech.reset();
    this.setState({ attempt: undefined, userAudioUrl: null }, () => {
      this.startSpeech();
    });
  }

  handleContinueClick() {
    const { items, currentIndex } = this.state;
    const item = items[currentIndex];
    const { attempt } = this.state;

    this.props.completeItem(item, attempt.mistakes);

    this.props.speech.reset();

    this.props.audio.stop();
    this.props.userAudio.stop();
    this.props.audio.cleanup();
    this.props.userAudio.cleanup();

    this.setState({ attempt: undefined, userAudioUrl: null });

    if (currentIndex < items.length - 1) {
      const nextIndex = currentIndex + 1;
      this.setState({
        currentIndex: nextIndex
      });
    } else {
      this.props.onFinish();
    }
  }

  handleAttempt(attempt) {
    const { text, solved } = attempt;

    this.setState({
      attempt: {
        text,
        mistakes: solved ? 0 : 2
      }
    });

    return new Promise((resolve) => {
      const options = {
        onEnded: resolve
      };
      playFeedbackSound(FEEDBACK_SOUND.forItemSolved(solved), options);
    });
  }

  startSpeech() {
    const { speech, learnLanguageBCP47, audio, recordingService } = this.props;

    if (audio.isPlaying) audio.stop();

    const { items, currentIndex } = this.state;
    const item = items[currentIndex];

    if (!(speech.listening || speech.recording)) {
      recordingService.start()
        .then((userAudioUrl) => {
          this.setState({ userAudioUrl });
        });

      speech.start({
        interimResults: true,
        lang: learnLanguageBCP47,
        continuous: true,
        targetText: item.learnLanguageText,
        itemUuid: item.id
      });
    }
  }

  handleSpeechError() {
    const { items, currentIndex } = this.state;
    const item = items[currentIndex];

    const { speech: { error, engineName } } = this.props;

    this.props.attemptItem(item, {
      solved: false,
      engineName,
      error
    });
  }

  handleSpeechEnd() {
    const {
      speech: {
        transcript,
        solved,
        error
      }
    } = this.props;

    this.props.recordingService.stop();

    if (!transcript || error) return;

    const { items, currentIndex } = this.state;
    const item = items[currentIndex];

    this.handleAttempt({ text: transcript, solved }).then(() => {
      const { audio, mediaUrlService } = this.props;
      const itemSoundUrl = mediaUrlService.soundURL(item.sound.id, 'mp3');
      // play only if the user is still on the same item
      if (solved && this.state.currentIndex === currentIndex) {
        audio.playSoundWithState(itemSoundUrl);
      }
    });

    this.props.attemptItem(item, {
      text: transcript,
      solved
    });
  }

  handleOnboardingComplete() {
    this.setState({
      onboardingCompleted: true
    });
    localStorage.set(SPEAKING_TRAINER_ONBOARDING_COMPLETE_KEY, true);
  }

  render() {
    if (this.state.currentIndex === 1 && !this.state.onboardingCompleted) {
      return <MotivationScreen onFinish={this.handleOnboardingComplete} />;
    }

    const { items, currentIndex } = this.state;
    const item = items[currentIndex];

    return (
      <SpeakingTrainer
        item={item}
        mediaUrlService={this.props.mediaUrlService}
        recordingService={this.props.recordingService}
        speech={this.props.speech}
        audio={this.props.audio}
        attempt={this.state.attempt}
        onMicButtonClick={this.handleMicButtonClick}
        onSoundPlayEnded={this.handleSoundPlayEnded}
        onContinue={this.handleContinueClick}
        onRetry={this.handleAttemptRetry}
        userAudioUrl={this.state.userAudioUrl}
        userAudio={this.props.userAudio}
        onFeedbackAttemptCardClick={this.handleFeedbackAttemptCardClick}
        onItemPlaybackCardClick={this.handleItemPlaybackCardClick}
      />
    );
  }
}

SpeakingTrainerContainer.propTypes = {
  learnLanguageAlpha3: PropTypes.string.isRequired,
  learnLanguageBCP47: PropTypes.string.isRequired,
  locale: PropTypes.string.isRequired,
  trainer: PropTypes.object.isRequired,
  onStart: PropTypes.func.isRequired,
  onFinish: PropTypes.func.isRequired,
  mediaUrlService: PropTypes.any.isRequired,
  recordingService: PropTypes.any.isRequired,
  speech: speechPropTypes.isRequired,
  track: PropTypes.func.isRequired,
  audio: audioPropTypes,
  userAudio: audioPropTypes,
  attemptItem: PropTypes.func.isRequired,
  completeItem: PropTypes.func.isRequired
};

export const mapStateToProps = ({ session, permissions }) => ({
  learnLanguageAlpha3: session.learnLanguageAlpha3,
  learnLanguageBCP47: learnLanguageAlpha3toBCP47(session.learnLanguageAlpha3),
  locale: session.locale,
  micPermission: permissions.micPermission
});

const mapDispatchToProps = {
  completeItem: sequenceActions.completeItem,
  attemptItem: sequenceActions.attemptItem,
  track
};

export default compose(
  withServices(['mediaUrlService', 'recordingService']),
  withAudio(),
  withAudio({ name: 'userAudio' }),
  withSpeech,
  withSpeechMessages,
  connect(mapStateToProps, mapDispatchToProps)
)(SpeakingTrainerContainer);
