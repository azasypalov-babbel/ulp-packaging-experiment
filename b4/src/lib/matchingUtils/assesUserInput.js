import { normalise } from './normalise';
import { evaluate } from './evaluate';
import { fetchSubstitutionCosts } from './fetchSubstitutionCosts';

export const assesUserInput = ({ attemptText, targetText, learnLanguageAlpha3 }) => {
  const normalisedAttemptText = normalise(attemptText, learnLanguageAlpha3);
  const normalisedTargetText = normalise(targetText, learnLanguageAlpha3);

  return fetchSubstitutionCosts(learnLanguageAlpha3)
    .then((substitutionCosts) =>
      evaluate({
        attemptText: normalisedAttemptText,
        targetText: normalisedTargetText,
        substitutionCosts
      })
    );
};
