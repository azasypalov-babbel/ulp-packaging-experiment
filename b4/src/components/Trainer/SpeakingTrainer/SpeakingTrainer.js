import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../shared/Text';
import { StyledWrapper } from './styles';
import Title from '../../shared/Title';
import SoundPlayer from '../../shared/SoundPlayer';
import { audioPropTypes } from '../../shared/withAudio';
import MicButton from '../../shared/MicButton/MicButton';
import { speechPropTypes } from '../../shared/withSpeech';
import ItemPlaybackCard from '../../shared/ItemPlaybackCard';
import FeedbackAttemptCard from '../../shared/FeedbackAttemptCard';
import FeedbackSheet from '../../shared/FeedbackSheet/FeedbackSheet';
import { replaceMatches } from '../../../lib/matchingUtils/normalise';
import StyledTrainerLayout from '../../TrainerLayout';
import pickRandom from '../../../lib/pickRandom';
import withTranslations from '../../shared/withTranslations';

const StyledUserInteractionWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const StyledMicButtonWrapper = styled.div`
  margin: 1rem;
`;

const isCorrectAttempt = (attempt) => attempt.mistakes < 1;

export class SpeakingTrainer extends Component {
  constructor() {
    super();
    this.renderSoundPlayer = this.renderSoundPlayer.bind(this);
    this.renderUserSoundPlayer = this.renderUserSoundPlayer.bind(this);
    this.handleSoundPlayEnded = this.handleSoundPlayEnded.bind(this);
    this.handleContinueClick = this.handleContinueClick.bind(this);
    this.state = {
      isUserAllowedToSpeak: false
    };
  }

  handleSoundPlayEnded() {
    const { isUserAllowedToSpeak } = this.state;
    if (!isUserAllowedToSpeak) {
      this.props.onSoundPlayEnded();
      this.setState({
        isUserAllowedToSpeak: true
      });
    }
  }

  handleContinueClick() {
    this.setState({
      isUserAllowedToSpeak: false
    }, () => this.props.onContinue());
  }

  renderSoundPlayer(props) {
    const { speech, recordingService, onItemPlaybackCardClick, item } = this.props;

    const triggerSound = () => {
      if (speech.listening || speech.recording) {
        recordingService.stop();
        speech.stop();
      }
      props.triggerSound();
      onItemPlaybackCardClick();
    };

    return (
      <ItemPlaybackCard
        {...item}
        {...props}
        onClick={triggerSound}
      />
    );
  }

  renderUserSoundPlayer(props) {
    const {
      speech: { transcript, lookupTableEntries },
      attempt,
      item,
      onFeedbackAttemptCardClick
    } = this.props;

    const triggerSound = () => {
      props.triggerSound();
      onFeedbackAttemptCardClick();
    };

    return (
      <FeedbackAttemptCard
        {...props}
        positive={isCorrectAttempt(attempt)}
        negative={!isCorrectAttempt(attempt)}
        onClick={triggerSound}
      >
        {isCorrectAttempt(attempt)
          ? item.learnLanguageText
          : replaceMatches(transcript, lookupTableEntries)}
      </FeedbackAttemptCard>
    );
  }

  renderFeedbackSheet() {
    const { onRetry, attempt, translations } = this.props;
    const isCorrect = isCorrectAttempt(attempt);

    const feedbackMessage = isCorrect
      ? translations.attemptFeedback.positiveFeedback
      : translations.attemptFeedback.negativeFeedback;
    const retryHandler = !isCorrect ? onRetry : null;

    return (
      <FeedbackSheet
        key="feedback-sheet"
        feedbackMessage={feedbackMessage}
        onContinueClick={this.handleContinueClick}
        onTryAgainClick={retryHandler}
      />
    );
  }

  render() {
    const {
      item,
      onMicButtonClick,
      attempt,
      speech: {
        listening,
        recording,
        ended
      },
      audio,
      userAudio,
      userAudioUrl,
      translations
    } = this.props;

    const showUserAttempt = attempt && userAudioUrl;

    const { isUserAllowedToSpeak } = this.state;

    const shouldRenderSpeakNowLabel = isUserAllowedToSpeak && (listening || recording);
    const shouldRenderClickToSpeakLabel = isUserAllowedToSpeak && !(listening || recording) && ended;

    return (
      <StyledTrainerLayout
        data-trainer-type="spokenreview"
        data-interaction-type="speak"
      >
        <StyledWrapper>
          <Title fontSize="big">{translations.trainerTitle}</Title>
          <StyledUserInteractionWrapper>
            <SoundPlayer
              playOnMount
              key={item.id}
              audio={audio}
              url={this.props.mediaUrlService.soundURL(item.sound.id)}
              render={this.renderSoundPlayer}
              onEnded={this.handleSoundPlayEnded}
              onError={this.handleSoundPlayEnded}
            />

            {showUserAttempt && (
              <SoundPlayer
                audio={userAudio}
                url={userAudioUrl}
                render={this.renderUserSoundPlayer}
              />
            )}
            {!attempt &&
              <>
                <StyledMicButtonWrapper>
                  <MicButton
                    isListening={listening}
                    isUserSpeaking={recording}
                    onClick={onMicButtonClick}
                    disabled={!isUserAllowedToSpeak}
                  />
                </StyledMicButtonWrapper>
                <Text fontFamily="fontMilliardMedium" color="spaceGrayW15" fontSize="base">
                  {shouldRenderSpeakNowLabel && translations.micButton.speak}
                  {shouldRenderClickToSpeakLabel && translations.micButton.pressToSpeak}
                </Text>
              </>}
          </StyledUserInteractionWrapper>
          {attempt && this.renderFeedbackSheet()}
        </StyledWrapper>
      </StyledTrainerLayout>
    );
  }
}

SpeakingTrainer.propTypes = {
  onRetry: PropTypes.func.isRequired,
  item: PropTypes.shape({
    id: PropTypes.string.isRequired,
    sound: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired,
    learnLanguageText: PropTypes.string.isRequired,
    displayLanguageText: PropTypes.string.isRequired
  }).isRequired,
  recordingService: PropTypes.any.isRequired,
  mediaUrlService: PropTypes.shape({
    soundURL: PropTypes.func.isRequired
  }).isRequired,
  onMicButtonClick: PropTypes.func,
  attempt: PropTypes.shape({
    mistakes: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired
  }),
  audio: audioPropTypes.isRequired,
  speech: speechPropTypes.isRequired,
  onContinue: PropTypes.func.isRequired,
  userAudioUrl: PropTypes.object,
  userAudio: audioPropTypes,
  translations: PropTypes.shape({
    micButton: PropTypes.shape({
      pressToSpeak: PropTypes.string.isRequired,
      speak: PropTypes.string.isRequired
    }).isRequired,
    trainerTitle: PropTypes.string.isRequired,
    attemptFeedback: PropTypes.shape({
      negativeFeedback: PropTypes.string.isRequired,
      positiveFeedback: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  onFeedbackAttemptCardClick: PropTypes.func.isRequired,
  onItemPlaybackCardClick: PropTypes.func.isRequired,
  onSoundPlayEnded: PropTypes.func
};

const getTranslations = (translate, props, instanceId) => {
  const positiveFeedbackMessages = [
    'speech_recognition.feedback.positive.1',
    'speech_recognition.feedback.positive.2',
    'speech_recognition.feedback.positive.3'
  ].map(translate);

  return {
    trainerTitle: translate('speech_recognition.trainer_item.title'),
    micButton: {
      pressToSpeak: translate('speech_recognition.mic_button.press_to_speak'),
      tapToSpeak: translate('speech_recognition.mic_button.tap_to_speak'),
      speak: `${translate('speech_recognition.mic_button.speak')}...`
    },
    attemptFeedback: {
      positiveFeedback: pickRandom(positiveFeedbackMessages, instanceId),
      negativeFeedback: translate('speech_recognition.feedback.negative.1')
    }
  };
};

SpeakingTrainer.defaultProps = {
  onSoundPlayEnded: () => { }
};

export default withTranslations(getTranslations)(SpeakingTrainer);
