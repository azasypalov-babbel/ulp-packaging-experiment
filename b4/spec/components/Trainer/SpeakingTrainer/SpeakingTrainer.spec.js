import React from 'react';
import { shallow } from 'enzyme';
import { SpeakingTrainer } from '../../../../src/components/Trainer/SpeakingTrainer/SpeakingTrainer';
import { replaceMatches } from '../../../../src/lib/matchingUtils/normalise';
jest.mock('../../../../src/lib/matchingUtils/normalise');
jest.mock('../../../../src/components/shared/FeedbackSheet/FeedbackSheet', () => 'FeedbackSheet');

describe('<SpeakingTrainer />', () => {
  afterEach(() => {
    replaceMatches.mockClear();
  });

  const audio = {
    playSound: jest.fn(),
    playSoundWithState: jest.fn(),
    preload: jest.fn(),
    reset: jest.fn(),
    stop: jest.fn()
  };

  const speech = {
    permissionsGranted: true,
    ended: false,
    error: '',
    listening: false,
    nomatch: false,
    recording: false,
    transcript: '',
    start: () => { },
    reset: () => { }
  };

  const defaultProps = {
    onContinue: () => { },
    onFeedbackAttemptCardClick: jest.fn(),
    onItemPlaybackCardClick: jest.fn(),
    onMicButtonClick: () => { },
    speech,
    mediaUrlService: {
      soundURL: jest.fn((url) => `mock-url-${url}.mp3`)
    },
    recordingService: {
      stop: jest.fn(),
      start: jest.fn(() => Promise.resolve())
    },
    item: {
      id: '123',
      sound: { id: '123' },
      learnLanguageText: 'learnLanguageText',
      displayLanguageText: 'displayLanguageText'
    },
    onRetry: jest.fn(),
    audio,
    translations: {
      micButton: {
        speak: 'Speak now',
        pressToSpeak: 'Press and start speaking'
      },
      trainerTitle: 'Listen, then say it out loud:',
      attemptFeedback: {
        labelYouSaid: 'We heard:',
        negativeFeedback: 'Youâ€™re getting there!',
        positiveFeedback: 'Sounds great!'
      }
    }
  };

  const defaultSoundPlayerProps = {
    triggerSound: jest.fn()
  };

  test('renders', () => {
    const wrapper = shallow(<SpeakingTrainer {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('onFeedbackAttemptCardClick callback', () => {
    it('should be called on trigger users recorded sound', () => {
      const attempt = {
        text: 'die Stadt',
        mistakes: 0
      };
      const wrapper = shallow(<SpeakingTrainer
        {...defaultProps}
        attempt={attempt}
      />);
      const userSoundPlayer = shallow(wrapper.instance().renderUserSoundPlayer(defaultSoundPlayerProps));
      const simulateAttemptCardClick = userSoundPlayer.first().prop('onClick');
      simulateAttemptCardClick();
      expect(defaultProps.onFeedbackAttemptCardClick).toHaveBeenCalled();
    });
  });

  describe('onItemPlaybackCardClick callback', () => {
    it('should be called on trigger of item sound playback', () => {
      const attempt = {
        text: 'die Stadt',
        mistakes: 0
      };
      const wrapper = shallow(<SpeakingTrainer
        {...defaultProps}
        attempt={attempt}
      />);
      const userSoundPlayer = shallow(wrapper.instance().renderSoundPlayer(defaultSoundPlayerProps));
      const simulateAttemptCardClick = userSoundPlayer.first().prop('onClick');
      simulateAttemptCardClick();
      expect(defaultProps.onItemPlaybackCardClick).toHaveBeenCalled();
    });
  });

  describe('when audio is done', () => {
    const onSoundPlayEnded = jest.fn();
    let component;
    beforeEach(() => {
      onSoundPlayEnded.mockClear();
      component = shallow(<SpeakingTrainer
        {...defaultProps}
        onSoundPlayEnded={onSoundPlayEnded}
      />);

      component.instance().handleSoundPlayEnded();
    });

    test('should call onSoundPlayEnded', () => {
      expect(onSoundPlayEnded).toHaveBeenCalled();
    });

    test('should set isUserAllowedToSpeak in state', () => {
      component.instance().handleSoundPlayEnded();
      const isUserAllowedToSpeak = component.state('isUserAllowedToSpeak');
      expect(isUserAllowedToSpeak).toBeTruthy();
    });

    test('should not call onSoundPlayEnded twice', () => {
      component.instance().handleSoundPlayEnded();
      expect(onSoundPlayEnded).toHaveBeenCalledTimes(1);
    });
  });

  describe('feedback attempt text', () => {
    describe('when correct attempt', () => {
      test('learnLanguageText is shown', () => {
        const attempt = {
          text: 'die Stadt',
          mistakes: 0
        };
        const item = {
          id: '123',
          sound: { id: '123' },
          learnLanguageText: 'die kleine Stadt',
          displayLanguageText: 'small city'
        };
        const wrapper = shallow(<SpeakingTrainer
          {...defaultProps}
          attempt={attempt}
          item={item}
        />);
        const userSoundPlayer = shallow(wrapper.instance().renderUserSoundPlayer(defaultSoundPlayerProps));
        expect(userSoundPlayer.contains('die kleine Stadt')).toEqual(true);
      });
      test('FeedbackSheet has one button', () => {
        const attempt = {
          text: 'die Stadt',
          mistakes: 0
        };
        const wrapper = shallow(<SpeakingTrainer
          {...defaultProps}
          attempt={attempt}
        />);

        expect(wrapper.find('FeedbackSheet')).toMatchSnapshot();
      });
    });

    describe('when incorrect attempt', () => {
      test('transcript is shown', () => {
        replaceMatches.mockImplementationOnce(() => 'die grosse Strasse');

        const attempt = {
          text: 'die Stadt',
          mistakes: 1
        };
        const item = {
          id: '123',
          sound: { id: '123' },
          learnLanguageText: 'die kleine Stadt',
          displayLanguageText: 'small city'
        };
        const wrapper = shallow(<SpeakingTrainer
          {...defaultProps}
          attempt={attempt}
          item={item}
          speech={{ ...speech, transcript: 'die grosse Strasse' }}
        />);
        const userSoundPlayer = shallow(wrapper.instance().renderUserSoundPlayer(defaultSoundPlayerProps));
        expect(replaceMatches).toHaveBeenCalledWith('die grosse Strasse', undefined);
        expect(userSoundPlayer.contains('die grosse Strasse')).toEqual(true);
      });
      test('FeedbackSheet has two buttons', () => {
        const attempt = {
          text: 'die Stadt',
          mistakes: 1
        };
        const wrapper = shallow(<SpeakingTrainer
          {...defaultProps}
          attempt={attempt}
        />);

        expect(wrapper.find('FeedbackSheet')).toMatchSnapshot();
      });
    });
  });
});
