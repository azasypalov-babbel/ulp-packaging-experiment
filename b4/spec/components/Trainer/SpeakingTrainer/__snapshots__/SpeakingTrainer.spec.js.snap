// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpeakingTrainer /> feedback attempt text when correct attempt FeedbackSheet has one button 1`] = `
<FeedbackSheet
  feedbackMessage="Sounds great!"
  key="feedback-sheet"
  onContinueClick={[Function]}
  onTryAgainClick={null}
/>
`;

exports[`<SpeakingTrainer /> feedback attempt text when incorrect attempt FeedbackSheet has two buttons 1`] = `
<FeedbackSheet
  feedbackMessage="Youâ€™re getting there!"
  key="feedback-sheet"
  onContinueClick={[Function]}
  onTryAgainClick={[MockFunction]}
/>
`;

exports[`<SpeakingTrainer /> renders 1`] = `
<styled.div
  data-interaction-type="speak"
  data-trainer-type="spokenreview"
>
  <styled.div>
    <Title
      as="h1"
      color="spaceGray"
      fontFamily="fontMilliardSemi"
      fontSize="big"
      textAlign="center"
    >
      Listen, then say it out loud:
    </Title>
    <styled.div>
      <SoundPlayer
        audio={
          Object {
            "playSound": [MockFunction],
            "playSoundWithState": [MockFunction],
            "preload": [MockFunction],
            "reset": [MockFunction],
            "stop": [MockFunction],
          }
        }
        key="123"
        onEnded={[Function]}
        onError={[Function]}
        playOnMount={true}
        render={[Function]}
        url="mock-url-123.mp3"
      />
      <styled.div>
        <MicButton
          disabled={true}
          isListening={false}
          isUserSpeaking={false}
          onClick={[Function]}
        />
      </styled.div>
      <styled.span
        color="spaceGrayW15"
        fontFamily="fontMilliardMedium"
        fontSize="base"
      />
    </styled.div>
  </styled.div>
</styled.div>
`;
